File: src/feishu/card_builder/project_cards.py
"""项目相关卡片构建模块

包含与项目管理相关的所有卡片构建函数：
- 项目列表卡片
- 项目信息卡片
"""

import logging
from datetime import datetime
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)


def build_project_list_card(
    projects: list,
    current_project_name: Optional[str] = None,
    confirming_project: Optional[str] = None,
) -> Dict[str, Any]:
    """
    构建带「切换」「删除」按钮的项目列表卡片（Schema 2.0 格式）

    当前项目用绿色高亮标识，非当前项目显示详情 + 切换按钮。
    点击按钮后飞书推送 im.card.action.trigger_v1 事件，handler 处理。

    Args:
        projects:             Project 对象列表
        current_project_name: 当前激活项目名（用于标记 ⭐）
        confirming_project:   处于二次确认删除状态的项目名

    Returns:
        飞书卡片 JSON（Schema 2.0，含 config/header/body/elements）
    """
    elements: List[Dict[str, Any]] = []

    if not projects:
        elements.append(
            {
                "tag": "markdown",
                "content": "📭 **暂无项目**\n\n使用 `/pa <路径>` 添加项目，或 `/pc <路径>` 创建新项目。",
            }
        )
    else:
        # ── 当前激活区块（绿色高亮）────────────────────────────────────────
        current_project = next(
            (p for p in projects if p.name == current_project_name), None
        )
        if current_project:
            p = current_project
            path = str(p.path)
            exists = p.exists()
            last_active = (
                p.last_active.strftime("%m-%d %H:%M")
                if isinstance(p.last_active, datetime)
                else str(p.last_active)[:16]
            )
            exists_hint = "✅ 目录存在" if exists else "⚠️ 目录不存在"

            elements.append(
                {
                    "tag": "markdown",
                    "content": "<font color='grey'>当前激活</font>",
                }
            )
            elements.append(
                {
                    "tag": "markdown",
                    "content": (
                        f"<font color='green'>🟢 **{p.display_name}**</font>\n"
                        f"标识: `{p.name}` · 活跃: {last_active}\n"
                        f"路径: `{path}`\n"
                        f"<font color='grey'>{exists_hint}</font>"
                    ),
                }
            )
            elements.append({"tag": "hr"})

        # ── 非激活项目：详情 + 操作按钮 ─────────────────────────────────────
        inactive_projects = [p for p in projects if p.name != current_project_name]
        for i, p in enumerate(inactive_projects, 1):
            name = p.name
            display_name = p.display_name
            path = str(p.path)
            exists = p.exists()
            is_confirming = name == confirming_project

            last_active = (
                p.last_active.strftime("%m-%d %H:%M")
                if isinstance(p.last_active, datetime)
                else str(p.last_active)[:16]
            )

            inactive_icon = "🟡" if exists else "🔴"
            not_exists_hint = (
                "\n<font color='red'>⚠️ 目录不存在</font>" if not exists else ""
            )
            elements.append(
                {
                    "tag": "markdown",
                    "content": (
                        f"{inactive_icon} **{display_name}**{not_exists_hint}\n"
                        f"标识: `{name}` · 活跃: {last_active}\n"
                        f"路径: `{path}`"
                    ),
                }
            )

            # 操作按钮
            if is_confirming:
                elements.append(
                    {
                        "tag": "button",
                        "text": {"tag": "plain_text", "content": "⚠️ 确认删除"},
                        "type": "danger",
                        "value": {
                            "action": "delete_project_confirmed",
                            "project_name": name,
                        },
                    }
                )
                elements.append(
                    {
                        "tag": "button",
                        "text": {"tag": "plain_text", "content": "取消"},
                        "type": "default",
                        "value": {
                            "action": "delete_project_cancel",
                            "project_name": name,
                        },
                    }
                )
            elif not exists:
                # 目录不存在：仅删除
                elements.append(
                    {
                        "tag": "button",
                        "text": {"tag": "plain_text", "content": "🗑️ 删除"},
                        "type": "danger",
                        "value": {
                            "action": "delete_project_confirm",
                            "project_name": name,
                        },
                    }
                )
            else:
                # 正常项目：切换 + 删除
                elements.append(
                    {
                        "tag": "button",
                        "text": {"tag": "plain_text", "content": "▶ 切换至此"},
                        "type": "primary",
                        "value": {
                            "action": "switch_project",
                            "project_name": name,
                        },
                    }
                )
                elements.append(
                    {
                        "tag": "button",
                        "text": {"tag": "plain_text", "content": "🗑️ 删除"},
                        "type": "danger",
                        "value": {
                            "action": "delete_project_confirm",
                            "project_name": name,
                        },
                    }
                )

            if i < len(inactive_projects):
                elements.append({"tag": "hr"})

    # ── 底部说明 ───────────────────────────────────────────────────────
    elements.append({"tag": "hr"})
    elements.append(
        {
            "tag": "markdown",
            "content": "📌 **命令说明**\n\n`/pa <路径>` — 添加已有目录为项目\n`/pc <路径>` — 创建新目录并添加项目",
        }
    )
    elements.append(
        {
            "tag": "markdown",
            "content": "<font color='grey'>💡 点击「🗑️ 删除」仅从列表移除，不会删除磁盘上的目录</font>",
        }
    )

    title_text = f"📁 项目列表（共 {len(projects)} 个）" if projects else "📁 项目列表"
    return {
        "schema": "2.0",
        "config": {"wide_screen_mode": True, "update_multi": True},
        "header": {
            "title": {"tag": "plain_text", "content": title_text},
            "template": "blue",
        },
        "body": {"elements": elements},
    }


def build_project_info_card(
    project: Any,
    is_current: bool = False,
) -> Dict[str, Any]:
    """
    构建项目信息详情卡片（Schema 2.0 格式）

    Args:
        project: Project 对象
        is_current: 是否为当前激活项目

    Returns:
        飞书卡片 JSON（Schema 2.0）
    """
    elements: List[Dict[str, Any]] = []

    # 当前项目标记
    active_marker = "🟢 **当前项目**\n\n" if is_current else ""
    status_icon = "✅" if project.exists() else "⚠️"
    status_text = "目录存在" if project.exists() else "目录不存在"

    # 项目基本信息
    elements.append(
        {
            "tag": "markdown",
            "content": (
                f"{active_marker}"
                f"**{project.display_name}**\n"
                f"<font color='grey'>标识: `{project.name}`</font>"
            ),
        }
    )
    elements.append({"tag": "hr"})

    # 详细信息
    created_at = (
        project.created_at.strftime("%Y-%m-%d %H:%M")
        if isinstance(project.created_at, datetime)
        else str(project.created_at)[:16]
    )
    last_active = (
        project.last_active.strftime("%Y-%m-%d %H:%M")
        if isinstance(project.last_active, datetime)
        else str(project.last_active)[:16]
    )

    info_lines = [
        f"**路径**: `{project.path}`",
        f"**状态**: {status_icon} {status_text}",
        f"**创建时间**: {created_at}",
        f"**最后活跃**: {last_active}",
        f"**关联会话**: {len(project.session_ids)} 个",
    ]
    if project.description:
        info_lines.append(f"**描述**: {project.description}")

    elements.append(
        {
            "tag": "markdown",
            "content": "\n".join(info_lines),
        }
    )

    # 操作按钮
    elements.append({"tag": "hr"})
    if not is_current and project.exists():
        elements.append(
            {
                "tag": "button",
                "text": {"tag": "plain_text", "content": "▶ 切换到此项目"},
                "type": "primary",
                "value": {
                    "action": "switch_project",
                    "project_name": project.name,
                },
            }
        )
    elements.append(
        {
            "tag": "button",
            "text": {"tag": "plain_text", "content": "📋 查看项目列表"},
            "type": "default",
            "value": {"action": "show_project_list"},
        }
    )

    return {
        "schema": "2.0",
        "config": {"wide_screen_mode": True, "update_multi": True},
        "header": {
            "title": {"tag": "plain_text", "content": "📁 项目信息"},
            "template": "blue",
        },
        "body": {"elements": elements},
    }

File: src/card-builder/project-cards.ts
/**
 * Project cards builder
 * 项目卡片构建模块
 *
 * 项目管理相关的卡片构建函数
 */

import type { CardElement, FeishuCard } from './base.js';

/** 项目信息 */
export interface ProjectInfo {
  name: string;
  displayName: string;
  path: string;
  exists: boolean;
  lastActive: Date;
  description?: string;
}

/**
 * 构建项目列表卡片
 *
 * @param projects - 项目列表
 * @param currentName - 当前项目名称
 * @param page - 当前页码（从1开始）
 * @param totalPages - 总页数
 * @param totalCount - 总项目数
 */
export function buildProjectListCard(
  projects: ProjectInfo[],
  currentName: string | null,
  page: number = 1,
  totalPages: number = 1,
  totalCount?: number
): FeishuCard {
  const elements: CardElement[] = [];

  // 分页配置：每页3个，最多12个（4页）
  const ITEMS_PER_PAGE = 3;
  const MAX_ITEMS = 12;
  const actualTotalCount = totalCount ?? projects.length;

  // 分页切片
  const currentPage = Math.max(1, Math.min(page, totalPages));
  const startIndex = (currentPage - 1) * ITEMS_PER_PAGE;
  const endIndex = Math.min(startIndex + ITEMS_PER_PAGE, Math.min(projects.length, MAX_ITEMS));
  const paginatedProjects = projects.slice(startIndex, endIndex);

  elements.push({
    tag: 'markdown',
    content: `📁 **项目列表** (${actualTotalCount} 个)`,
  });
  elements.push({ tag: 'hr' });

  if (projects.length === 0) {
    elements.push({
      tag: 'markdown',
      content: '暂无项目，使用 `/pa <路径>` 添加已有项目，或 `/pc <路径>` 创建新项目。',
    });
  } else {
    for (let i = 0; i < paginatedProjects.length; i++) {
      const p = paginatedProjects[i];
      const marker = p.name === currentName ? ' ★' : '';
      const status = p.exists ? '✅' : '❌';

      elements.push({
        tag: 'markdown',
        content: `**${startIndex + i + 1}.** ${status} **${p.displayName}**${marker}`,
      });
      elements.push({
        tag: 'markdown',
        content: `标识: \`${p.name}\`\n路径: \`${p.path}\`\n活跃: ${p.lastActive.toLocaleString('zh-CN')}`,
      });

      if (p.description) {
        elements.push({
          tag: 'markdown',
          content: `描述: ${p.description}`,
        });
      }

      if (i < paginatedProjects.length - 1) {
        elements.push({ tag: 'hr' });
      }
    }

    // 分页控件
    if (totalPages > 1) {
      elements.push({ tag: 'hr' });
      elements.push({
        tag: 'column_set',
        flex_mode: 'none',
        columns: [
          {
            tag: 'column',
            width: 'auto',
            elements: [{
              tag: 'button',
              text: { tag: 'plain_text', content: '⬅️ 上一页' },
              type: 'default',
              value: { action: 'project_page', page: currentPage - 1 },
              disabled: currentPage <= 1,
            }],
          },
          {
            tag: 'column',
            width: 'weighted',
            weight: 1,
            vertical_align: 'center',
            elements: [{
              tag: 'markdown',
              content: `**第 ${currentPage}/${totalPages} 页**`,
              text_align: 'center',
            }],
          },
          {
            tag: 'column',
            width: 'auto',
            elements: [{
              tag: 'button',
              text: { tag: 'plain_text', content: '下一页 ➡️' },
              type: 'default',
              value: { action: 'project_page', page: currentPage + 1 },
              disabled: currentPage >= totalPages,
            }],
          },
        ],
      });
    }
  }

  elements.push({ tag: 'hr' });
  elements.push({
    tag: 'markdown',
    content: "💡 `/ps <标识>` 切换项目 · `/pi <标识>` 查看详情",
  });

  return {
    schema: '2.0',
    header: {
      title: { tag: 'plain_text', content: '📁 项目管理' },
      template: 'blue',
    },
    body: { elements },
  };
}

/**
 * 构建项目详情卡片
 */
export function buildProjectInfoCard(
  project: ProjectInfo,
  isCurrent: boolean
): FeishuCard {
  const elements: CardElement[] = [];

  if (isCurrent) {
    elements.push({
      tag: 'markdown',
      content: '<font color="green">🟢 当前激活项目</font>',
    });
  }

  elements.push({
    tag: 'markdown',
    content: `**${project.displayName}**\n\n标识: \`${project.name}\`\n路径: \`${project.path}\``,
  });

  if (project.description) {
    elements.push({
      tag: 'markdown',
      content: `描述: ${project.description}`,
    });
  }

  elements.push({
    tag: 'markdown',
    content: `状态: ${project.exists ? '✅ 目录存在' : '❌ 目录不存在'}\n最后活跃: ${project.lastActive.toLocaleString('zh-CN')}`,
  });

  return {
    schema: '2.0',
    header: {
      title: { tag: 'plain_text', content: '📁 项目详情' },
      template: isCurrent ? 'green' : 'blue',
    },
    body: { elements },
  };
}
